truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1365,309 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
+        @Child private TaintNode taintNode = TaintNode.create();
+        @Child private CallDispatchHeadNode strCompatAndValidNode = DispatchHeadNodeFactory.createMethodCall(true);
+        @Child private RopeNodes.MakeConcatNode makeConcatNode = RopeNodesFactory.MakeConcatNodeGen.create(null, null, null);
+
+        @Specialization
+        public DynamicObject scrubNoBlock(VirtualFrame frame, DynamicObject string, Object repl, NotProvided block,
+                                          @Cached("createBinaryProfile()") ConditionProfile validRangeProfile,
+                                          @Cached("createBinaryProfile()") ConditionProfile asciiCompatibleProfile) {
+            return scrubDefault(frame, string, repl, nil(), validRangeProfile, asciiCompatibleProfile);
+        }
+
+        @Specialization
+        public DynamicObject scrubDefault(VirtualFrame frame, DynamicObject string, Object repl, DynamicObject block,
+                                          @Cached("createBinaryProfile()") ConditionProfile validRangeProfile,
+                                          @Cached("createBinaryProfile()") ConditionProfile asciiCompatibleProfile) {
+            final DynamicObject rubyEncoding = getContext().getEncodingManager().getRubyEncoding(StringOperations.encoding(string));
+            final Rope rope = rope(string);
+            final ByteList value = RopeOperations.getByteListReadOnly(rope);
+
+            CodeRange cr = rope.getCodeRange();
+            Encoding enc;
+            Encoding encidx;
+            Rope buf = null;
+            byte[] repBytes;
+            int rep;
+            int replen;
+            boolean tainted = false;
+
+            if (validRangeProfile.profile(cr == CR_7BIT || cr == CR_VALID)) {
+                return nil();
+            }
+
+            enc = rope.getEncoding();
+            if (repl != nil()) {
+                repl = strCompatAndValidNode.call(frame, string, "str_compat_and_valid", repl, rubyEncoding);
+                tainted |= isTaintedNode.executeIsTainted(repl);
+            }
+
+            if (enc.isDummy()) {
+                return nil();
+            }
+            encidx = enc;
+
+            if (asciiCompatibleProfile.profile(enc.isAsciiCompatible())) {
+                byte[] pBytes = value.unsafeBytes();
+                int p = value.begin();
+                int e = p + value.getRealSize();
+                int p1 = p;
+                boolean rep7bit_p;
+                if (block != nil()) {
+                    repBytes = null;
+                    rep = 0;
+                    replen = 0;
+                    rep7bit_p = false;
+                }
+                else if (repl != nil()) {
+                    final Rope replRope = rope((DynamicObject) repl);
+                    repBytes = replRope.getBytesCopy();
+                    rep = 0;
+                    replen = repBytes.length;
```

Comment:
@eregon I've cleaned up these variables and made some final in the latest commit.